freeze state in signalState

rainerhahnekamp opened the pull request:
It is currently possible to run mutable changes via the `$update` function. Those mutable changes do not throw any error, and effects tracking that part of the state are also not executed. The tests are located in the "immutability" tests suite of **signal-state.spec.ts**.

This change applies `Object.freeze` to the state, causing an error on mutable changes.

The implementation is an almost identical copy of `freeze` in NgRx.

Other than that,
- move the tests to the test directory, following the existing conventions.
- add `withEffect` as a test utility method to simplify the tests of effects.

rainerhahnekamp committed:
freeze state in signalState

rainerhahnekamp committed:
only freeze state in signal state, if the state has been updated with a function, i.e. state is exposed.

@markostanimirovic: I am not happy with the `@ts-ignore` in `deepFreeze`. Maybe you can find a better way.

markostanimirovic commented on src/lib/signal-state.ts:
```diff
@@ -16,6 +17,7 @@ export type SignalStateUpdate<State extends Record<string, unknown>> = {
 export function signalState<State extends Record<string, unknown>>(
   initialState: State
 ): SignalState<State> {
+  deepFreeze(initialState);
```
we should perform `deepFreeze` only in dev mode for better performance

rainerhahnekamp replied:
Still working on that

markostanimirovic commented on src/lib/utils.ts:
```diff
@@ -0,0 +1,40 @@
+export function deepFreeze<T extends object>(target: T): T {
```
let's name this file `deep-freeze.ts` and keep `isFunction` and `isObjectLike` functions private

rainerhahnekamp replied:
Done

markostanimirovic commented on src/lib/utils.ts:
```diff
@@ -0,0 +1,40 @@
+        ? prop !== 'caller' && prop !== 'callee' && prop !== 'arguments'
+        : true)
+    ) {
+      // @ts-ignore
+      const propValue = target[prop];
```
will `(target as any)[prop]` help here to get rid of the `// @ts-ignore`? 👀

rainerhahnekamp replied:
I went with a Record<string, unknown>. What is better in that case?

markostanimirovic commented on src/tests/test2.ts:
```diff
@@ -1,5 +1,4 @@
 import {
-  selectSignal,
```
`test1.ts` and `test2.ts` files contain "type tests 😅", to make sure that typing works fine after refactoring. When we move this lib to the NgRx repo, they will be converted into real type tests. For now, they contain a few commented lines that use `selectSignal` so feel free to revert this change.

rainerhahnekamp replied:
done. I've left testEffects there as it is more a testing file and not a test

markostanimirovic commented on src/tests/with-effect.ts:
```diff
@@ -0,0 +1,26 @@
+ * withEffects renders a TestComponent and executes the tests inside an injectionContext.
+ * This allows us to generate effects very easily.
+ */
+export const withEffect =
```
I'd rename this function to avoid potential naming confusion with SignalStore features:

```suggestion
export const testEffect =
```

rainerhahnekamp replied:
done

rainerhahnekamp committed:
- rename util to deep-freeze
- rename withEffect to testEffects
- type assertion in freezing function
- move signal state spec back to original location

rainerhahnekamp committed:
fix tests

rainerhahnekamp committed:
freeze signalState optionally (only in dev mode). $update function provides state as readonly

- refactored tests
- added further tests for mutable and immutable changes

markostanimirovic commented:
Have you tried to run the playground app? It seems that the `signalStore` typing is broken. 👀 

If you agree, we can keep only the `deepFreeze` related changes in this PR and consider introducing `DeepReadonly` and other typing changes separately.

Regarding the `{ immutabilityCheck: boolean }` option in the `signalState` function, I don't see the use-case when it should be set to `false` because `signalState` (and `signalStore`) support only immutable updates. Therefore, I suggest keeping the same signature as before (without `options` argument) and calling `deepFreeze` in dev mode:

```ts
if (isDevMode()) {
  deepFreeze(state);
}
```

---

Btw, `@ngrx/store` provides the ability to turn off the state immutability check, but the main reason for that was to ensure backward compatibility when runtime checks were introduced.

rainerhahnekamp commented:
Yeah, you're right. Let's get that freeze feature in and discuss the rest in a separate PR.
